File: solutions/scala/matching-brackets/src/main/scala/MatchingBrackets.scala
import scala.annotation.tailrec

object MatchingBrackets:
  private val BracketPairs = Map(
    ']' -> '[', ')' -> '(', '}' -> '{'
  )
  private val OpeningBrackets = BracketPairs.values.toSet
  private val ClosingBrackets = BracketPairs.keySet

  @tailrec
  def isPaired(brackets: String, stack: List[Char] = List()): Boolean =
    if brackets.isEmpty then stack.isEmpty
    else if ClosingBrackets contains brackets.head
    then stack match
        case openingBracket :: remainingStack
          if BracketPairs(brackets.head) == openingBracket =>
          isPaired(brackets.tail, remainingStack)
        case _ => false
    else if OpeningBrackets contains brackets.head
    then isPaired(brackets.tail, brackets.head :: stack)
    else isPaired(brackets.tail, stack)

File: solutions/python/matching-brackets/matching_brackets.py
_OPENING_BRACKETS = "({["
_CLOSING_BRACKETS = ")}]"
_BRACKETS = _OPENING_BRACKETS + _CLOSING_BRACKETS

def is_matching(a, b):
    return _OPENING_BRACKETS.index(a) == _CLOSING_BRACKETS.index(b)

def is_paired(input_string: str) -> bool:
    brackets = (c for c in input_string if c in _BRACKETS)
    stack = []
    for bracket in brackets:
        if bracket in _OPENING_BRACKETS:
            stack.append(bracket)
        elif not stack or not is_matching(stack.pop(), bracket):
            return False

    return not stack
